docs/Developers/Design/GD2_node_state_machine.md

```diff
@@ -0,0 +1,131 @@
+While the duties of the operator and GD2 are largely disjoint, there is some
+level of coordination required between them. The operator is responsible for
+adding and removing resources (nodes) from the GD2-managed cluster, and GD2 is
+responsible for managing the free capacity and volumes within that cluster.
+
+The main area for coordination is when removing nodes from the cluster and when
+performing maintenance (that may result in downtime of a node).
+
+Concretely:
+
+- When removing a node from the cluster, GD2 should have the opportunity to
+  remove data from that node prior to its deletion. While not strictly
+  necessary, this allows proactive migration of data to maintain appropriate
+  data redundancy.
+- When a node may be offline for maintenance, it is necessary to notify GD2
+  such that it does not decide the node has permanently failed. Such permanent
+  failure designation would lead to needlessly rebuilding data to other
+  locations in the cluster.
+
+To facilitate this coordination, the following state machine is proposed. Note
+that the diagram is just an approximation to keep the diagram simple, while
+also providing a useful overview of the states and transitions.
+
+![state machine diagram](state_machine.dot.png)
+
+The above state machine would apply independently to each node of the Gluster
+cluster (i.e., each node has its own *current state*). The high-level meaning
+of each state is:
+
```

Comment:
Is this state maintained by operator ? Or do you expect glusterd2 to maintain this ?

Reply:
Per the suggestion farther down in this doc, the idea is to have a set of bool flags that would be maintained by GD2 per-node. Those flags are visible to both operator and GD2, and (depending on the specific flag) be modified by one or both of operator or gd2. See below for flag list and who is allowed to modify.